pkg/compactor/blocks_cleaner_test.go

```diff
@@ -1266,104 +1267,180 @@ func TestBlocksCleaner_RaceCondition_CleanerUpdatesBucketIndexWhileAnotherCleane
+								})
+							}
+						}
+					}(),
+				}
 
-	// Create two blocks and mark one of them for deletion at a time before the deletion delay.
-	block1 := createTSDBBlock(t, bucketClient, tenantID, 10, 20, 1, nil)
-	block2 := createTSDBBlock(t, bucketClient, tenantID, 20, 30, 1, nil)
-	createDeletionMark(t, bucketClient, tenantID, block1, time.Now().Add(-deletionDelay).Add(-time.Minute))
-
-	// Control the bucket clients in order to reproduce the following scenario:
-	// the 2nd cleaner lists markers after the 1st cleaner has started deleting some blocks.
-	cleaner2ListMarkersStarted := make(chan struct{})
-	cleaner2ListMarkersUnblocked := make(chan struct{})
-
-	cleaner1BucketClient := &hookBucket{
-		Bucket: bucketClient,
-		preIterHook: func(_ context.Context, dir string, _ func(string) error, _ ...objstore.IterOption) {
-			// When listing blocks, wait until the 2nd cleaner started to list markers.
-			if path.Base(dir) == tenantID {
-				<-cleaner2ListMarkersStarted
-			}
+				cleaner2BucketClient = &hookBucket{
+					Bucket: bucketClient,
+					preIterHook: func() func(context.Context, string, func(string) error, ...objstore.IterOption) {
+						once := sync.Once{}
+
+						return func(_ context.Context, dir string, _ func(string) error, _ ...objstore.IterOption) {
+							if path.Base(dir) == block.MarkersPathname {
+								// When listing markers, signal that the 2nd cleaner started to list markers and
+								// then wait until it should proceed.
+								once.Do(func() {
+									close(cleaner2ListMarkersStarted)
+									<-cleaner2ListMarkersUnblocked
+								})
+							}
+						}
+					}(),
+				}
+
+				return
+			},
 		},
-		preDeleteHook: func() func(context.Context, string) {
-			once := sync.Once{}
-
-			return func(_ context.Context, name string) {
-				// When deleting the deletion mark of a block (which is expected to be the last object deleted of a block)
-				// unblock the 2nd cleaner markers listing with some delay. The delay is used to make sure this deletion
-				// completes before the 2nd cleaner start listing markers.
-				if path.Base(name) == block.DeletionMarkFilename {
-					once.Do(func() {
-						go func() {
-							time.Sleep(100 * time.Millisecond)
-							close(cleaner2ListMarkersUnblocked)
-						}()
-					})
+		"the 2nd cleaner fetches new deletion marks after the 1st cleaner has started deleting blocks for the new deletion marks": {
+			mockBucketClients: func(bucketClient objstore.Bucket) (cleaner1BucketClient, cleaner2BucketClient objstore.Bucket) {
+				cleaner2ListMarkersStarted := make(chan struct{})
+				cleaner2GetDeletionMarkUnblocked := make(chan struct{})
+
+				// Mock the 1st cleaner to delete deletion marks when the 2nd cleaner is between the listing
+				// of deletion marks and fetching their content (for new markers).
+				cleaner1BucketClient = &hookBucket{
+					Bucket: bucketClient,
+					preDeleteHook: func() func(context.Context, string) {
+						once := sync.Once{}
```

Comment:
Isn't this a new `Once` instance every time the `preDeleteHook` is called? I don't understand how it's different from just calling the func below upon every invocation of `preDeleteHook`. Having said that, I still think a `postDeleteHook` is better.

Reply:
The `Once` is global because a channel can only be closed once, so we protect from it. However, in this specific test we only expect 1 block to be deleted, so it shouldn't matter much.

Reply:
I had missed that it's a `func() func(...)`, so the closure with the once is reused for every call of `Delete()` on the client. SGTM